### ISSUE TITLE
[FEA] Remove `object` dtype

### ISSUE THREAD

**Issue by brandon-b-miller** [rapidsai/cudf#5714]
**Is your feature request related to a problem? Please describe.**
The `np.object` data type is used within cuDF python to represent string data, but this doesn't really mirror the purpose or function of `object` datatypes in CPU based python libraries, particularly numpy and pandas which can work with truly general python objects. This results in a lot of special handling in certain places in the python code as well as giving the wrong impression to users about our handling of python objects. We should replace it with something better, possibly an extension type similar to list/categorical in the short term. 

**Describe the solution you'd like**
Replace `np.object` in the cuDF codebase with a `cudf.StringDtype` object.

**Describe alternatives you've considered**
We could also try and shoehorn `pd.StringDtype()` in instead or wait and handle this as part of a wider `cud.Dtype` effort.

**Comment by GregoryKimball** [comment]
@brandon-b-miller I like your proposal. Seeing "object" in my string dtypes confuses me every time I see it.

**Comment by vyasr** [comment]
@mroeschke this is probably something we want to address as part of pandas 3.0 upgrade, right?

**Comment by mroeschke** [comment]
Yup, superseding this issue with https://github.com/rapidsai/cudf/issues/20818
